File: python/behavioral/17_memento.py
"""
设计模式: 备忘录模式 (Memento)
分类: 行为型

意图: 在不破坏封装性的前提下，捕获对象内部状态，以便日后恢复。
核心思想: 发起人创建/恢复备忘录，管理者保存历史，三角色分离。
编程范式: OOP — 状态快照；Python 用 dataclass + 嵌套类保护封装。
"""

from __future__ import annotations
from dataclasses import dataclass
from collections import deque


# ── 发起人 ────────────────────────────────────────────────────────────────
class Editor:

    @dataclass(frozen=True)
    class Memento:
        """不透明快照，仅 Editor 应使用"""
        _content: str

    def __init__(self) -> None:
        self._content = ""

    def type(self, text: str) -> None:
        self._content += text

    def print(self) -> None:
        print(f'Content: "{self._content}"')

    def save(self) -> "Editor.Memento":
        return Editor.Memento(self._content)

    def restore(self, memento: "Editor.Memento") -> None:
        self._content = memento._content


# ── 管理者 ────────────────────────────────────────────────────────────────
class History:
    def __init__(self) -> None:
        self._stack: deque[Editor.Memento] = deque()

    def push(self, m: Editor.Memento) -> None: self._stack.append(m)
    def pop(self) -> Editor.Memento | None:
        return self._stack.pop() if self._stack else None


# ── 演示 ──────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    editor  = Editor()
    history = History()

    editor.type("Hello");           history.push(editor.save())
    editor.type(", World");         history.push(editor.save())
    editor.type("!!!");             editor.print()   # Hello, World!!!

    editor.restore(history.pop()); editor.print()   # Hello, World
    editor.restore(history.pop()); editor.print()   # Hello

File: cpp/behavioral/17_memento.cpp
/**
 * 设计模式: 备忘录模式 (Memento)
 * 分类: 行为型
 *
 * 意图: 在不破坏封装性的前提下，捕获一个对象的内部状态，以便日后恢复。
 * 核心思想: 发起人创建/恢复备忘录，管理者负责保存备忘录列表，三角色分离。
 * 编程范式: OOP — 状态快照 + 封装保护（发起人私有状态不对外暴露）。
 */

#include <iostream>
#include <memory>
#include <stack>
#include <string>

// ── 备忘录（不透明令牌）──────────────────────────────────────────────────
class Memento {
    friend class Editor;   // 只有 Editor 可访问内部
    std::string state_;
    explicit Memento(std::string s) : state_(std::move(s)) {}
};

// ── 发起人 ────────────────────────────────────────────────────────────────
class Editor {
    std::string content_;
public:
    void type(const std::string& text) { content_ += text; }
    void print() const { std::cout << "Content: \"" << content_ << "\"\n"; }

    std::unique_ptr<Memento> save() const {
        return std::unique_ptr<Memento>(new Memento(content_));
    }
    void restore(const Memento& m) { content_ = m.state_; }
};

// ── 管理者 ────────────────────────────────────────────────────────────────
class History {
    std::stack<std::unique_ptr<Memento>> stack_;
public:
    void push(std::unique_ptr<Memento> m) { stack_.push(std::move(m)); }
    std::unique_ptr<Memento> pop() {
        if (stack_.empty()) return nullptr;
        auto m = std::move(stack_.top());
        stack_.pop();
        return m;
    }
};

int main() {
    Editor  editor;
    History history;

    editor.type("Hello");
    history.push(editor.save());   // 保存快照 1

    editor.type(", World");
    history.push(editor.save());   // 保存快照 2

    editor.type("!!!");
    editor.print();                // Hello, World!!!

    auto m2 = history.pop();
    editor.restore(*m2);
    editor.print();                // Hello, World

    auto m1 = history.pop();
    editor.restore(*m1);
    editor.print();                // Hello
    return 0;
}
